### PR TITLE
Add PHPStan to Magento static checks

### PR DESCRIPTION
**Pull request by lenaorobei**
## Problem

Following types of issues can be found in Magento codebase:
- the use of undefined class;
- call to undefined function;
- missing `parent::__construct()` call;
- passed arguments mismatch, etc.

Currently such bugs cannot be detected by tools Magento uses for static checks.
[PHPStan](https://github.com/phpstan/phpstan) utilizes AST (abstract syntax tree) to perform sophisticated static code analysis which cannot be archived by using token-base approach (PHP CodeSniffer) and Reflection (PHP Mess Detector).

## Solution

Add `phpstan/phpstan` to Magento `require-dev` section, update static code checks.

## Requested Reviewers

@buskamuza 
@paliarush 
@antonkril

### CODE REVIEW

**Comment by magento-cicd2** [comment]
[CLA assistant check](https://cla.dev.magento.com/magento/architecture?pullRequest=114) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dev.magento.com/magento/architecture?pullRequest=114) before we can accept your contribution.<br/>

**Comment by larsroettig** [comment]
We can use to bootstrap 
- https://github.com/fooman/phpstan-magento2-magic-methods
- https://github.com/fooman/magento2-dev-bootstrap

**Comment by shochdoerfer** [comment]
If you need or want to validate .phtml files also, you might need to follow this [blog post](http://martens.me/programming/phpstan-static-code-analyzer-for-magento-2.html). Not sure if patching phpstan is needed or if one could simply write an extension.

File: design-documents/testing/static/phpstan.md [hunk 1]
```diff
@@ -0,0 +1,103 @@
+# Overview
+[PHPStan](https://github.com/phpstan/phpstan) is a code analysis tool which focuses on finding bugs without running the code.
+PHPStan uses [nikic/PHP-Parser](https://github.com/nikic/PHP-Parser) under the hood to build AST (abstract syntax tree) and perform sophisticated static code analysis.
+
+### Prerequisites
+- PHP >= 7.1.
+- Installed Magento application with generated code.
```
Comment (buskamuza) [inline design-documents/testing/static/phpstan.md]: Can we avoid installed instance? Is it only for generated code?

Comment (lenaorobei) [reply design-documents/testing/static/phpstan.md]: Yes, we can. My bad, wrong wording. `setup:di:compile` is good enough.

**Comment by shochdoerfer** [comment]
The Drupal people also picked up phpstan, read some insights [here](https://glamanate.com/blog/writing-better-drupal-code-static-analysis-using-phpstan).

**Pull request merged**